src/software/thunderscope/robot_diagnostics/error_log_widgets.py

```diff
@@ -0,0 +1,273 @@
+        self.info_layout = QVBoxLayout()
+        self.info_layout.setContentsMargins(
+            0, self.icon_size / 4, 0, self.icon_size / 3
+        )
+
+        # robot id displayed more prominently
+        self.robot_id = QLabel(f"Robot {robot_id}")
+        self.robot_id.setStyleSheet(
+            "font-weight: bold;" f"font-size: {int(self.height() / 25)}px;"
+        )
+
+        # message of the error
+        self.message = QLabel(message)
+        self.message.setStyleSheet(f"font-size: {int(self.height() / 35)}px;")
+
+        self.info_layout.addWidget(self.robot_id)
+        self.info_layout.addWidget(self.message)
+
+        # layout for timestamp info
+        self.time_layout = QVBoxLayout()
+        self.time_layout.setContentsMargins(
+            0, self.icon_size / 4, 0, self.icon_size / 3
+        )
+
+        # the timestamp the error took place
+        self.timestamp = timestamp if timestamp else datetime.now()
+        self.timestamp_label = QLabel(self.timestamp.strftime("%H:%M:%S %f"))
+        # time since the error was logged (easier to understand than timestamp)
+        self.time_since_label = QLabel(str(0))
+
+        self.time_layout.addWidget(self.timestamp_label)
+        self.time_layout.addWidget(self.time_since_label)
+
+        # adding spacing to look nice
+        self.layout.addWidget(self.icon)
+        self.layout.addStretch(1)
+        self.layout.addLayout(self.info_layout)
+        self.layout.addStretch(16)
+        self.layout.addLayout(self.time_layout)
+        self.layout.setContentsMargins(0, 0, 0, 0)
+
+        self.setLayout(self.layout)
+        self.base_stylesheet = textwrap.dedent(
+            f"""
+            RobotLogMessageWidget {{
+                border: 0;
+                padding: {self.icon_size / 22};
+            }}
+            """
+        )
+        self.setStyleSheet(self.base_stylesheet)
+
+    def update_time(self):
+        """
+        Updates the time since label for the error log message
+        """
+        # get the time difference from the timestamp to now
+        time_diff = datetime.now() - self.timestamp
+        self.time_since_label.setText(f"{int(time_diff.total_seconds())} seconds ago")
+
+    def mouseReleaseEvent(self, event):
+        """
+        Mouse Release Event handler for child classes to override
+        :param event: the mouse release event
+        """
+        super().mouseReleaseEvent(event)
+
+
+class LowBatteryLogMessageWidget(RobotLogMessageWidget):
+    """
+    A robot error log message when a robot has low battery
+    """
+
+    def __init__(self, robot_id):
+        super(LowBatteryLogMessageWidget, self).__init__(
+            robot_id, "Battery voltage is low", error_constants.get_low_battery_icon()
+        )
+
+
+class ErrorCodeLogMessageWidget(RobotLogMessageWidget):
+    """
+    A robot error log message when it has an error code
+    """
+
+    def __init__(self, robot_id: int, error_code: str):
```

Comment:
All `__init__` methods in this class seem to be missing their docstring

Reply:
done